### PR TITLE
New pattern - Empty States

### PR DESCRIPTION
**Pull request by kamebkj**
**Overview**
This PR includes a new pattern component - Empty States, with a few examples and test cases. The documentation for Empty State will be in a separate PR later once this component is in the Evergreen package.

**Screenshots**
- At a high level, an Empty State consists of a logo, some text, and the actions to be taken.
- There are two types of Empty States: `horizontal` (default) and `vertical`, which can be specified with `orientation` prop. 
- Additionally, when using an Empty State in a table, we should use the `light` background, vs the `dark` background when it is not in a table (does not have a border container) and is standalone, by specifying the `background` prop.

Here's a basic example of what a standalone, horizontal Empty State looks like:
image

An example of what it looks like in a table:
image

An example of what it looks like in a smaller viewport such as a popover:
image

### CODE REVIEW

File: src/empty-states/__tests__/EmptyState.test.js
Comment (akleiner2) [inline src/empty-states/__tests__/EmptyState.test.js:30]: What do you think about updating assertions to use `toBeVisible()`? It also includes `toBeInTheDocument()` as implied.

File: src/empty-states/__tests__/EmptyState.test.js
Comment (akleiner2) [inline src/empty-states/__tests__/EmptyState.test.js:22]: Let's update this test to check for the presence of the icon / image

File: src/empty-states/__tests__/EmptyState.test.js
Comment (akleiner2) [inline src/empty-states/__tests__/EmptyState.test.js:84]: We can get rid of this now :)

File: src/empty-states/src/EmptyState.js
Comment (akleiner2) [inline src/empty-states/src/EmptyState.js:15]: Should we name these props `icon` and `iconBgColor`?

File: src/empty-states/src/EmptyState.js
Comment (akleiner2) [inline src/empty-states/src/EmptyState.js:134]: One thing that I'm thinking about...do we know / can we guarantee that the icon color and iconbackground color can be compatible? This puts responsibility on consumers to "do the right thing", but I'm just wondering if there's a way we can restrict this API up front

File: src/empty-states/src/EmptyState.js
Comment (akleiner2) [inline src/empty-states/src/EmptyState.js:127]: The way this currently behaves - we're assuming links passed in would strictly be external links. Should we remove the explicit `target="_blank"`, and instead leave that up to consumers? I can see folks passing through single-page route changes here... i.e.


```
<EmptyState 
    // ... 
    anchorCta={<EmptyState.LinkButton is={ReactRouterLink} to="/path/in/my-app"}
/>
```

File: src/empty-states/src/EmptyState.js
Comment (kamebkj) [reply src/empty-states/src/EmptyState.js:15]: Yup good call!

File: src/empty-states/src/EmptyState.js
Comment (kamebkj) [reply src/empty-states/src/EmptyState.js:134]: Yeah agree that this is putting the responsibility to the consumers. A few ideas:
- Restricted the icon color & background color with intent (success, warn, danger, info) and only allow passing in icon shape (class) and its intent. Not sure if this would be too restricted.
- Restricted to just icon and its color, and we always apply the color minus 200 (red-500 -> red-300) as its background so we get one less prop. This can result in less optimized contrast since our color system was not great.
- Keep as is.

What do you think?

Comment (akleiner2) [reply src/empty-states/src/EmptyState.js]: Hm..yeah I say we can keep as-is for now and maybe work to make the API more restrictive in the future.

File: src/empty-states/src/EmptyState.js
Comment (akleiner2) [inline src/empty-states/src/EmptyState.js:11]: Let's memoize these components to prevent unnecessary re-renders. 

You can do: 

```
import React, { memo } from 'react' 
```

and then `const HoriztontalOrientation = memo(function HorizontalOrientation({ ... }) { return ... })`

File: src/empty-states/src/EmptyState.js
Comment (akleiner2) [inline src/empty-states/src/EmptyState.js:77]: `memo` here.

File: src/empty-states/src/EmptyState.js
Comment (akleiner2) [inline src/empty-states/src/EmptyState.js:130]: memo here :)

**Comment by akleiner2** [comment]
This is looking pretty good! Left a few more small things to touch up. The final thing is to add this component export to the bundle, and add its types to `index.d.ts`. Then, we can add documentation after this lands + we release.

**Review by akleiner2 — APPROVED** [review]
Woo!! Nice work - I think this is an excellent first ship :)

**Pull request merged**
